src/libcore/iter/mod.rs

```diff
@@ -737,6 +733,76 @@ impl<I> Iterator for StepBy<I> where I: Iterator {
     }
 }
 
+// hidden trait for specializing iterator methods
+// current use is limited to next() on StepBy
+trait SpecIter {
+    type Item;
+    fn next(&mut self) -> Option<Self::Item>;
+}
+
+impl<I> SpecIter for StepBy<I>
+where
+    I: Iterator,
+{
+    type Item = I::Item;
+
+    #[inline]
+    default fn next(&mut self) -> Option<I::Item> {
+        if self.first_take {
+            self.first_take = false;
+            self.iter.next()
+        } else {
+            self.iter.nth(self.step)
+        }
+    }
+}
+
+impl<T> SpecIter for StepBy<ops::Range<T>>
+where
+    T: Step,
+{
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        self.first_take = false;
+        if !(self.iter.start < self.iter.end) {
```

Comment:
This comparison is structured weirdly to handle NaN properly?

Reply:
yes, it's `if start < end { return stuff } else { None }` flipped around so the positive case has no indent.